lib/rex/socket/parameters.rb

```diff
@@ -231,6 +237,41 @@ def initialize(hash)
     self.v6        = hash['IPv6'] || false
   end
 
+  def to_hash
+    settings = {}
+    settings['PeerHost'] = self.peerhost unless self.peerhost.nil?
+    settings['LocalHost'] = self.localhost
+    settings['PeerPort'] = self.peerport
+    settings['LocalPort'] = self.localport
+    settings['Bare'] = self.bare
+
+    settings['SSL'] = self.ssl
+    settings['SSLContext'] = self.sslctx unless self.sslctx.nil?
+    settings['SSLVersion'] = self.ssl_version unless self.ssl_version.nil?
+    settings['SSLVerifyMode'] = self.ssl_verify_mode unless self.ssl_verify_mode.nil?
+    settings['SSLCompression'] = self.ssl_compression unless self.ssl_compression.nil?
+    settings['SSLCipher'] = self.ssl_cipher unless self.ssl_cipher.nil?
+    settings['SSLCommonName'] = self.ssl_cn unless self.ssl_cn.nil?
+    settings['SSLCertRaw'] = self.ssl_cert unless self.ssl_cert.nil?
+    settings['SSLClientCertRaw'] = self.ssl_client_cert unless self.ssl_client_cert.nil?
+    settings['SSLClientKeyRaw'] = self.ssl_client_key unless self.ssl_client_key.nil?
+
+    settings['Proxies'] = self.proxies.map{|a| a.join(':')}.join(',') unless self.proxies.nil?
+    settings['Proto'] = self.proto
+    settings['Server'] = self.server
+    settings['Comm'] = self.comm
+    settings['Context'] = self.context
+    settings['Retries'] = self.retries
+    settings['Timeout'] = self.timeout
+    settings['IPv6'] = self.v6
+
+    settings
+  end
+
+  def merge_hash(other_hash)
```

Comment:
Serialization is nice, but I think this would be better as a plain `merge` method that may or may not use hashes under the covers. Is there an anticipated need for `to_hash` aside from this method? If not, I would consider moving most of the code in `initialize` to something like `import_settings_from_hash` and have this method go like:
```ruby
def merge(other)
  case other
  when Hash
    import_settings_from_hash(other)
  when Parameters
    # import straight across if needed, possibly using metaprogramming
    other.instance_variables.each do |var|
      val = other.instance_variable_get(var)
      instance_variable_set(var, val) unless val.nil?
    end
  end
end
```
To accommodate this, the new `import_settings_from_hash` method would need to start checking for `self.xxxxx` when setting defaults, but I think that still might be better than maintaining the list of params and their hash keys in two different spots. Alternately, you could try untangling the bulk import from the default setting code, possibly moving some of the defaults into manually written accessor methods.

Reply:
Mostly, I feel having two separate hash key -> variable methods this large is a recipe for bugs. I'm open to other alternatives.

Reply:
My concern with that approach was handling the few settings that have defaults that are not nil like localhost, localport and peerport, ssl. We could handle those seperately but I'm not sure how we'd track the difference between explicitly setting localport to 0 and a localport that was left at it's default value of 0 and thus shouldn't be updated.

Reply:
That's where the custom accessors could come into play:
```ruby
attr_writer :localport
def localport
  @localport || 0
end
```
Doing the instance variable merge would then not overwrite things with default values from unset variables. This approach would also help with the serialize-deserialize approach currently here, which has the same issue.